x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,538 @@
+	gvg, found := k.GetGVG(ctx, gvgID)
+	if !found {
+		return nil, types.ErrGVGNotExist
+	}
+
+	// check staking
+	if gvg.StoredSize+expectedStoreSize > k.GetTotalStakingStoreSize(ctx, gvg) {
+		return nil, types.ErrInsufficientStaking.Wrapf("gvg state: %s", gvg.String())
+	}
+	return gvg, nil
+}
+
+func (k Keeper) SetSwapOutInfo(ctx sdk.Context, gvgFamilyID uint32, gvgIDs []uint32, spID uint32, successorSPID uint32) error {
+	store := ctx.KVStore(k.storeKey)
+
+	if gvgFamilyID != types.NoSpecifiedFamilyId {
+		key := types.GetSwapOutFamilyKey(gvgFamilyID)
+		found := store.Has(key)
+		if found {
+			return types.ErrSwapOutFailed.Wrapf("SwapOutInfo of this gvg family(ID: %d) already exist", gvgFamilyID)
+
+		}
+		swapOutInfo := &types.SwapOutInfo{
+			SpId:          spID,
+			SuccessorSpId: successorSPID,
+		}
+
+		bz := k.cdc.MustMarshal(swapOutInfo)
+		store.Set(key, bz)
+	} else {
+		for _, gvgID := range gvgIDs {
+			key := types.GetSwapOutGVGKey(gvgID)
+			found := store.Has(key)
+			if found {
+				return types.ErrSwapOutFailed.Wrapf("SwapOutInfo of this gvg(ID: %d) already exist.", gvgID)
+			}
+			swapOutInfo := &types.SwapOutInfo{
+				SpId:          spID,
+				SuccessorSpId: successorSPID,
+			}
+			bz := k.cdc.MustMarshal(swapOutInfo)
+			store.Set(key, bz)
+		}
+	}
+	return nil
+}
+
+func (k Keeper) DeleteSwapOutInfo(ctx sdk.Context, gvgFamilyID uint32, gvgIDs []uint32, spID uint32) error {
+	store := ctx.KVStore(k.storeKey)
+
+	swapOutInfo := types.SwapOutInfo{}
+	if gvgFamilyID != types.NoSpecifiedFamilyId {
+		key := types.GetSwapOutFamilyKey(gvgFamilyID)
+		bz := store.Get(key)
+		k.cdc.MustUnmarshal(bz, &swapOutInfo)
+
+		if swapOutInfo.SpId != spID {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("spID(%d) is different from the spID(%d) in swapOutInfo", spID, swapOutInfo.SpId)
+		} else {
+			store.Delete(key)
+		}
+
+	} else {
+		for _, gvgID := range gvgIDs {
+			key := types.GetSwapOutGVGKey(gvgID)
+			bz := store.Get(key)
+			k.cdc.MustUnmarshal(bz, &swapOutInfo)
+			if swapOutInfo.SpId != spID {
+				return sptypes.ErrStorageProviderNotFound.Wrapf("spID(%d) is different from the spID(%d) in swapOutInfo", spID, swapOutInfo.SpId)
+			} else {
+				store.Delete(key)
+			}
+		}
+	}
+	return nil
+}
+
+func (k Keeper) CompleteSwapOut(ctx sdk.Context, gvgFamilyID uint32, gvgIDs []uint32, successorSP *sptypes.StorageProvider) error {
+	store := ctx.KVStore(k.storeKey)
+
+	swapOutInfo := types.SwapOutInfo{}
+	if gvgFamilyID != types.NoSpecifiedFamilyId {
+		key := types.GetSwapOutFamilyKey(gvgFamilyID)
+		bz := store.Get(key)
+		k.cdc.MustUnmarshal(bz, &swapOutInfo)
+
+		sp, found := k.spKeeper.GetStorageProvider(ctx, swapOutInfo.SpId)
+		if !found {
+			return sptypes.ErrStorageProviderNotFound.Wrapf("The storage provider(ID: %d) not found when complete swap out.", swapOutInfo.SpId)
+		}
+
+		err := k.SwapOutAsPrimarySP(ctx, sp, successorSP, gvgFamilyID)
```

Comment:
I think we should check `SuccessorSpId` in the `SwapOutInfo` match here.

Reply:
Fixed